trin-bridge/src/pandaops.rs

```diff
@@ -0,0 +1,135 @@
+use crate::bridge::Retry;
+use crate::full_header::{FullHeader, FullHeaderBatch};
+use anyhow::{anyhow, bail};
+use ethereum_types::H256;
+use ethportal_api::types::jsonrpc::params::Params;
+use ethportal_api::types::jsonrpc::request::JsonRequest;
+use ethportal_api::utils::bytes::hex_encode;
+use ethportal_api::{Header, Receipts};
+use serde_json::{json, Value};
+
+pub struct PandaOpsMiddleware {
+    pub base_el_endpoint: String,
+    pub base_cl_endpoint: String,
+    pub client_id: String,
+    pub client_secret: String,
+}
+
+impl PandaOpsMiddleware {
+    pub fn new(
+        base_el_endpoint: String,
+        base_cl_endpoint: String,
+        client_id: String,
+        client_secret: String,
+    ) -> Self {
+        Self {
+            base_el_endpoint,
+            base_cl_endpoint,
+            client_id,
+            client_secret,
+        }
+    }
+}
+
+impl PandaOpsMiddleware {
+    /// Used the "surf" library here instead of "ureq" since "surf" is much more capable of handling
+    /// multiple async requests. Using "ureq" consistently resulted in errors as soon as the number of
+    /// concurrent tasks increased significantly.
+    async fn batch_request(&self, obj: Vec<JsonRequest>) -> anyhow::Result<String> {
+        let result = surf::post(self.base_el_endpoint.clone())
+            .middleware(Retry::default())
+            .body_json(&json!(obj))
+            .map_err(|e| anyhow!("Unable to construct json post request: {e:?}"))?
+            .header("Content-Type", "application/json".to_string())
+            .header("CF-Access-Client-Id", self.client_id.clone())
+            .header("CF-Access-Client-Secret", self.client_secret.clone())
+            .recv_string()
+            .await;
+
+        result.map_err(|_| anyhow!("Unable to request execution batch from pandaops"))
+    }
+
+    #[allow(dead_code)]
+    pub(crate) async fn request(&self, endpoint: String) -> anyhow::Result<String> {
+        let result = surf::get(endpoint)
+            .header("Content-Type", "application/json".to_string())
+            .header("CF-Access-Client-Id", self.client_id.clone())
+            .header("CF-Access-Client-Secret", self.client_secret.clone())
+            .recv_string()
+            .await;
+        result.map_err(|_| anyhow!("Unable to request consensus block from pandaops"))
+    }
+
+    pub(crate) async fn get_trusted_receipts(
```

Comment:
OK... I have to ask bc I've seen this before but couldn't find in the docs what's going on. What exactly is `pub(crate)` doing? Is this only a public function within this crate? What's your reasoning behind using this vs a pure `pub async fn...`?

Reply:
Yes, it is a public function within the crate and because this function is not used outside of the `trin-bridge` crate, it makes more sense to be public only for the crate. We can change this to pure public func if we need it in the future imo.